### PR TITLE
refactor(web): subkey abstraction (embedded step 3) - encapsulates subkey lookup, cancellation

### PR DESCRIPTION
**Pull request by jahorton**
For this next step (actually, _pair_ of steps), we now break the control flow for subkeys into two separate steps:

1. a "pending longpress"
    - This part is responsible for detecting and recognizing a complete longpress gesture, which then results in the display of a  subkey menu.
    - The gesture may be 'cancelled' or 'fulfilled', the latter of which results in display of the menu.
2. a "realized longpress" -> display of the subkey popup
    - Note that the `VisualKeyboard` class has code conditioned upon whether or not a subkey popup is visible in one form or other.
        - This was the role of the old `VisualKeyboard.popupVisible`, though just for the embedded codepath.

You might see where this is going in the arc overview - `PendingGesture` and `RealizedGesture`.

We would like to better encapsulate the code path into an object-oriented form that may be used to configure the OSK within the upcoming OSK-core module if possible.  The OSK-core won't have access to kmwembedded.ts, but there is a significant amount of its code that may be relocated into something that can be extracted and included within that module.

As it turns out, this is something a "two birds with one stone" scenario.  By determining a good encapsulation for the embedded code path that can also be hosted by the OSK-core, we can clarify the embedded control flow and identify its critical aspects.

------

Arc overview:

1. #5294
    - Step 1 theme:  subkey commands as async events (as _gestures_ are inherently async)
2. #5295
3. #5354
    - Step 2 theme:  code path cleanup / clarification
4. #5355
5. #5356
    - Step 3 theme:  prep for common abstraction for the two control flows
6. #5357
    - that's right - "Gesture".  These aren't intended for _just_ longpresses, though they'll be the only supported gesture at first.
    - As gestures are events that evaluate over intervals of time, they are inherently asynchronous.  Hence, the use of `Promise`s by earlier PRs in the chain.
7. #5358
    - Given a "gesture" abstraction, establishes gesture-oriented patterns 

A rough breakdown of some of the design goals + patterns may be found in the description of #5294.

### CODE REVIEW

File: web/source/osk/embedded/subkeyDelegator.ts [hunk 1]
```diff
@@ -1,16 +1,53 @@
...
 
-    public resolve(keyEvent: text.KeyEvent) {
+    public resolve(keyCoreID: string) {
       if(this.resolver) {
+        let keyEvent: text.KeyEvent = null;
+
+        if(keyCoreID != null) {
+          // This is set with the base key of our current subkey elsewhere within the engine.
+          var baseKey: OSKKeySpec = this.baseKey.key.spec;
+          var found = false;
```
Comment (mcdurdin) [inline web/source/osk/embedded/subkeyDelegator.ts:27]:
```suggestion
          let found;
```

`found` is always set below, just like `selectedKey`.

File: web/source/osk/embedded/subkeyDelegator.ts [hunk 2]
```diff
@@ -1,16 +1,53 @@
...
     }
 
-    public resolve(keyEvent: text.KeyEvent) {
+    public resolve(keyCoreID: string) {
       if(this.resolver) {
+        let keyEvent: text.KeyEvent = null;
+
+        if(keyCoreID != null) {
+          // This is set with the base key of our current subkey elsewhere within the engine.
+          var baseKey: OSKKeySpec = this.baseKey.key.spec;
```
Comment (mcdurdin) [inline web/source/osk/embedded/subkeyDelegator.ts:26]:
```suggestion
          let baseKey: OSKKeySpec = this.baseKey.key.spec;
```

_let_'s use consistent variable declarations in new code 🤣

File: web/source/osk/embedded/subkeyDelegator.ts [hunk 3]
```diff
@@ -1,16 +1,53 @@
...
+            found = true;
+          } else {
+            // ... yeah, there are some funky type shenanigans between the two.
+            // OSKKeySpec is the OSK's... reinterpretation of the ActiveKey type.
+            selectedKey = (baseKey as keyboards.ActiveKey).getSubkey(keyCoreID) as OSKKeySpec;
+            found = !!selectedKey;
+          }
+
+          if(!found) {
+            this.resolver(null); // Maintains existing behavior.
```
Comment (mcdurdin) [inline web/source/osk/embedded/subkeyDelegator.ts:41]: What does this comment mean? What existing behaviour?

File: web/source/osk/banner.ts [hunk 4]
```diff
@@ -493,7 +493,7 @@ namespace com.keyman.osk {
       // Utilized by the mobile apps; allows them to 'take over' touch handling,
       // blocking it within KMW when the apps are already managing an ongoing touch-hold.
       let keyman = com.keyman.singleton;
-      return keyman['osk'].vkbd.popupVisible;
+      return keyman['osk'].vkbd.subkeyDelegator;
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/banner.ts:496]: Good point re: "truthy".

As for the other point... probably.  Was trying to keep collateral down and make the change in a manner consistent with other replacements of `.popupVisible` to facilitate the review.

**Comment by jahorton** [comment]
Abandoning in favor of #5387 - a documented + combined version of all seven, as many of the changes in later steps reshuffle code from earlier steps.

**Pull request closed without merging**